Sugarchain: add: Yespower Cache PoW hashes

volbil opened the pull request:
# Originary implemented by 

https://github.com/ResistancePlatform/resistance-core/commit/a73d69465c8884e6289ecbdec58331e8fdc0376a
https://github.com/ResistancePlatform/resistance-core/commit/3c80dac3618e42c90057da7ba6fa49c44cf64349
https://github.com/ResistancePlatform/resistance-core/compare/20c7ba33101133053bd001a84479ff341a134cba...3c80dac3618e42c90057da7ba6fa49c44cf64349

-----

# Result

- IBD `3x` faster `7:30`
![image](https://user-images.githubusercontent.com/37016180/71312484-e0847100-246e-11ea-9d64-aff3a4c23acb.png)

- reindex `2.4x` faster `9:00`
![image](https://user-images.githubusercontent.com/37016180/71312493-ec703300-246e-11ea-9a51-4aa7d6270361.png)

-----

# Testing
- [x] rename: `GetPoWHashCached()` to `GetPoWHash_cached()` by original author (trivial)
- [x] make check: OK
- [x] test_bitcoin: OK
- [x] IBD: OK: took `7h30m at height 1858419` ***300% faster***
  * [x] UNCACHED: IBD: OK: took `22h20m at height 1874686`
- [x] `-reindex` OK: took `3h15m+3h25m==6h40m at height 1814177` ***200% faster***
  * [x] UNCACHED: `-reindex` OK: took `13h20m (6:40+6:40) at height 1843981`
- [x] `-reindex-chainstate` OK: took `3h40m at height 1809089` ***170% faster***
  * [x] UNCACHED `-reindex-chainstate` OK: took `6h13m at height 1828078`
- [ ] import bootstrap
- [ ] benchmark vs non-cache (at least 3 times)
  * [x] IBD
  * [x] reindex
  * [ ] reindex-chainstate

-----

# Log

```
X block = cba1cfff0fd1bc43ba5508f3113502d1f77c803afd9ca295e1ad2303ba80397e PoW = 000004e5ef92746b970adde07576e89f8f73434efc67cb7704fd0021bdcb5265
O block = cba1cfff0fd1bc43ba5508f3113502d1f77c803afd9ca295e1ad2303ba80397e PoW = 000004e5ef92746b970adde07576e89f8f73434efc67cb7704fd0021bdcb5265
X block = a3f0556b4db92179d5bf961f9c16406886a669b522caf43a18d6ad48b0b7490d PoW = 00000167626529c9ce87bb85e40f8f992359fb31ebdc9af49dbf1cb7ef2db6a1
O block = a3f0556b4db92179d5bf961f9c16406886a669b522caf43a18d6ad48b0b7490d PoW = 00000167626529c9ce87bb85e40f8f992359fb31ebdc9af49dbf1cb7ef2db6a1
X block = ed8461b33a2534f537d7a07e95c6dd9fa7f9915c3761ebd16f4cebc562e8b14c PoW = 000006b5c8d01e40596a914d0235d79bc9d4402e4c6665353440bc02b60697aa
O block = ed8461b33a2534f537d7a07e95c6dd9fa7f9915c3761ebd16f4cebc562e8b14c PoW = 000006b5c8d01e40596a914d0235d79bc9d4402e4c6665353440bc02b60697aa
X block = b32e64b7f26282eb2bf4990e4f03b83b6644ad71af3919aa89c5f87687f463c8 PoW = 0000004271c1f17d3f4c96bad6d69360b19bde4a9fdeacae01767768f7591ec9
O block = b32e64b7f26282eb2bf4990e4f03b83b6644ad71af3919aa89c5f87687f463c8 PoW = 0000004271c1f17d3f4c96bad6d69360b19bde4a9fdeacae01767768f7591ec9
X block = 031bc4827cec6b1b56292e870278b4d103150c51634f22d954fced9ec78dd153 PoW = 0000079b8815fe58bc73c99d60240b69e4aefa8623c4314599efbfb7af02aaa5
O block = 031bc4827cec6b1b56292e870278b4d103150c51634f22d954fced9ec78dd153 PoW = 0000079b8815fe58bc73c99d60240b69e4aefa8623c4314599efbfb7af02aaa5
X block = c7d37106b96c8e9e11d57a3b06ace1a3535cac972b7672602f98e1befad5db08 PoW = 000006323eab4101ef2445c6b8fdb713e11f9cfbb0be8d9a35fd1a878eeaf9b1
O block = c7d37106b96c8e9e11d57a3b06ace1a3535cac972b7672602f98e1befad5db08 PoW = 000006323eab4101ef2445c6b8fdb713e11f9cfbb0be8d9a35fd1a878eeaf9b1
X block = 34960049b2238e9ea079726ea713b21891984effedd5fc3be8e11bf55f7f8d88 PoW = 000002f51a41a53258bfc5a8b2659cc69da4ee931ecec7e4717833d09930c1e3
O block = 34960049b2238e9ea079726ea713b21891984effedd5fc3be8e11bf55f7f8d88 PoW = 000002f51a41a53258bfc5a8b2659cc69da4ee931ecec7e4717833d09930c1e3
X block = 2208b3e7de6591904c3a658d378f8a2713287919b736e9a0d1a1ed007dab76ee PoW = 00000430a393799e7db231169536868cf3a9146716039374c1e738a8b2e875fa
O block = 2208b3e7de6591904c3a658d378f8a2713287919b736e9a0d1a1ed007dab76ee PoW = 00000430a393799e7db231169536868cf3a9146716039374c1e738a8b2e875fa
X block = 8060d16d884d5631b012ff95507838b973a6c8be80b99ee3854d81a3867e666d PoW = 000006f24bc50a78842bd7c65691520ac2e263c64c3c799d1d81246d8b4707a7
O block = 8060d16d884d5631b012ff95507838b973a6c8be80b99ee3854d81a3867e666d PoW = 000006f24bc50a78842bd7c65691520ac2e263c64c3c799d1d81246d8b4707a7
X block = fab57cbb173d49f17e24b9672aa854300300cd8d3939d5562db66ba3f9aea19b PoW = 00000476e2a983b0ba617d5c290c469638760f676248c65ffeb7d8714213b31f
O block = fab57cbb173d49f17e24b9672aa854300300cd8d3939d5562db66ba3f9aea19b PoW = 00000476e2a983b0ba617d5c290c469638760f676248c65ffeb7d8714213b31f
X block = 3a8f9c58dcfd397fede17e9dc7fd23c0667b26a847ca3bbccddf6d567358fa7c PoW = 0000085dc5a994da2d4cc161ad4110d720043c553665bee67a25166a2cbfe134
O block = 3a8f9c58dcfd397fede17e9dc7fd23c0667b26a847ca3bbccddf6d567358fa7c PoW = 0000085dc5a994da2d4cc161ad4110d720043c553665bee67a25166a2cbfe134
```

- Sync Log
https://youtu.be/MojeGzy2NHU

-----

# TODO
- [ ] remove: `printf` in `block.cpp`

volbil committed:
Added work caching

cryptozeny commented:
how to test?

<details><summary>CLICK ME</summary>
<p>

## Basic Test
```
git clone git@github.com:sugarchain-project/sugarchain.git
cd sugarchain
git fetch origin pull/22/head:volbil-yespower_cache
git checkout volbil-yespower_cache
./autogen
./configure
make -j8
make check -j8
./src/test/test_sugarchain test_bitcoin --log_level=test_suite
```

## Synctime test
https://github.com/sugarchain-project/sugarchain/pull/26

</p>
</details>

cryptozeny commented:
no error found 8b450cc
```
make cheak

NO ERROR
```

cryptozeny commented:
8b450cc
```
./src/test/test_sugarchain test_bitcoin --log_level=test_suite
Running 282 test cases...

*** No errors detected
```

cryptozeny commented:
Testing TODO
- [x] rename: `GetPoWHashCached()` to `GetPoWHash_cached()` by original author (trivial)
- [x] make check: OK
- [x] test_bitcoin: OK
- [x] IBD: OK: took `7h30m at height 1858419` ***300% faster***
  * [x] UNCACHED: IBD: OK: took `22h20m at height 1874686`
- [x] `-reindex` OK: took `3h15m+3h25m==6h40m at height 1814177` ***200% faster***
  * [x] UNCACHED: `-reindex` OK: took `13h20m (6:40+6:40) at height 1843981`
- [x] `-reindex-chainstate` OK: took `3h40m at height 1809089` ***170% faster***
  * [x] UNCACHED `-reindex-chainstate` OK: took `6h13m at height 1828078`
- [ ] import bootstrap
- [ ] benchmark vs non-cache (at least 3 times)
  * [x] IBD
  * [x] reindex
  * [ ] reindex-chainstate

volbil commented:
Backported address index from this commits:
https://github.com/MicroBitcoinOrg/MicroBitcoin/commit/34ef71b734d4995592ad3a4c176bf2a81257e217
https://github.com/MicroBitcoinOrg/MicroBitcoin/commit/2ec3a1b5efe23a9848c2a7ad46b6258e6eed88b4

cryptozeny commented:
**DO NOT MERGE YET**

cryptozeny commented:
https://github.com/sugarchain-project/sugarchain/pull/26
benchmark

Commit:
Sugarchain: Yespower Cache: follow solardiz
switch: variable names after solardiz
add: logging yespower cache
add: comments
add: copyright: Alexander Peslyak

volbil committed:
Merge pull request #1 from sugarchain-project/volbil-cleanup-zeny

Yespower Cache: solardiz cleanup

cryptozeny commented on src/primitives/block.cpp:
```diff
@@ -42,21 +44,25 @@ uint256 CBlockHeaderUncached::GetPoWHash() const
             exit(1);
         }
+        // yespower cache: log
+        printf("HIT block_hash = %s PoW_hash = %s\n", cache_block_hash.ToString().c_str(), cache_PoW_hash.ToString().c_str());
```
check log vs RES, and if its OK, comment this out.

cryptozeny commented on src/primitives/block.cpp:
```diff
@@ -42,21 +44,25 @@ uint256 CBlockHeaderUncached::GetPoWHash() const
+        cache_block_hash = block_hash;
+        cache_init = true;
+        // yespower cache: log
+        printf("MISS block_hash = %s PoW_hash = %s\n", cache_block_hash.ToString().c_str(), cache_PoW_hash.ToString().c_str());
```
check log vs RES, and if its OK, comment this out.

cryptozeny commented:
8e5b110 no error on tests

cryptozeny commented:
sync logging
https://youtu.be/MojeGzy2NHU

IBD 3x faster `7:30`
![image](https://user-images.githubusercontent.com/37016180/71312484-e0847100-246e-11ea-9d64-aff3a4c23acb.png)

reindex 2.4x faster `9:00`
![image](https://user-images.githubusercontent.com/37016180/71312493-ec703300-246e-11ea-9a51-4aa7d6270361.png)

cryptozeny commented:
log
```
X block = cba1cfff0fd1bc43ba5508f3113502d1f77c803afd9ca295e1ad2303ba80397e PoW = 000004e5ef92746b970adde07576e89f8f73434efc67cb7704fd0021bdcb5265
O block = cba1cfff0fd1bc43ba5508f3113502d1f77c803afd9ca295e1ad2303ba80397e PoW = 000004e5ef92746b970adde07576e89f8f73434efc67cb7704fd0021bdcb5265
X block = a3f0556b4db92179d5bf961f9c16406886a669b522caf43a18d6ad48b0b7490d PoW = 00000167626529c9ce87bb85e40f8f992359fb31ebdc9af49dbf1cb7ef2db6a1
O block = a3f0556b4db92179d5bf961f9c16406886a669b522caf43a18d6ad48b0b7490d PoW = 00000167626529c9ce87bb85e40f8f992359fb31ebdc9af49dbf1cb7ef2db6a1
X block = ed8461b33a2534f537d7a07e95c6dd9fa7f9915c3761ebd16f4cebc562e8b14c PoW = 000006b5c8d01e40596a914d0235d79bc9d4402e4c6665353440bc02b60697aa
O block = ed8461b33a2534f537d7a07e95c6dd9fa7f9915c3761ebd16f4cebc562e8b14c PoW = 000006b5c8d01e40596a914d0235d79bc9d4402e4c6665353440bc02b60697aa
X block = b32e64b7f26282eb2bf4990e4f03b83b6644ad71af3919aa89c5f87687f463c8 PoW = 0000004271c1f17d3f4c96bad6d69360b19bde4a9fdeacae01767768f7591ec9
O block = b32e64b7f26282eb2bf4990e4f03b83b6644ad71af3919aa89c5f87687f463c8 PoW = 0000004271c1f17d3f4c96bad6d69360b19bde4a9fdeacae01767768f7591ec9
X block = 031bc4827cec6b1b56292e870278b4d103150c51634f22d954fced9ec78dd153 PoW = 0000079b8815fe58bc73c99d60240b69e4aefa8623c4314599efbfb7af02aaa5
O block = 031bc4827cec6b1b56292e870278b4d103150c51634f22d954fced9ec78dd153 PoW = 0000079b8815fe58bc73c99d60240b69e4aefa8623c4314599efbfb7af02aaa5
X block = c7d37106b96c8e9e11d57a3b06ace1a3535cac972b7672602f98e1befad5db08 PoW = 000006323eab4101ef2445c6b8fdb713e11f9cfbb0be8d9a35fd1a878eeaf9b1
O block = c7d37106b96c8e9e11d57a3b06ace1a3535cac972b7672602f98e1befad5db08 PoW = 000006323eab4101ef2445c6b8fdb713e11f9cfbb0be8d9a35fd1a878eeaf9b1
X block = 34960049b2238e9ea079726ea713b21891984effedd5fc3be8e11bf55f7f8d88 PoW = 000002f51a41a53258bfc5a8b2659cc69da4ee931ecec7e4717833d09930c1e3
O block = 34960049b2238e9ea079726ea713b21891984effedd5fc3be8e11bf55f7f8d88 PoW = 000002f51a41a53258bfc5a8b2659cc69da4ee931ecec7e4717833d09930c1e3
X block = 2208b3e7de6591904c3a658d378f8a2713287919b736e9a0d1a1ed007dab76ee PoW = 00000430a393799e7db231169536868cf3a9146716039374c1e738a8b2e875fa
O block = 2208b3e7de6591904c3a658d378f8a2713287919b736e9a0d1a1ed007dab76ee PoW = 00000430a393799e7db231169536868cf3a9146716039374c1e738a8b2e875fa
X block = 8060d16d884d5631b012ff95507838b973a6c8be80b99ee3854d81a3867e666d PoW = 000006f24bc50a78842bd7c65691520ac2e263c64c3c799d1d81246d8b4707a7
O block = 8060d16d884d5631b012ff95507838b973a6c8be80b99ee3854d81a3867e666d PoW = 000006f24bc50a78842bd7c65691520ac2e263c64c3c799d1d81246d8b4707a7
X block = fab57cbb173d49f17e24b9672aa854300300cd8d3939d5562db66ba3f9aea19b PoW = 00000476e2a983b0ba617d5c290c469638760f676248c65ffeb7d8714213b31f
O block = fab57cbb173d49f17e24b9672aa854300300cd8d3939d5562db66ba3f9aea19b PoW = 00000476e2a983b0ba617d5c290c469638760f676248c65ffeb7d8714213b31f
X block = 3a8f9c58dcfd397fede17e9dc7fd23c0667b26a847ca3bbccddf6d567358fa7c PoW = 0000085dc5a994da2d4cc161ad4110d720043c553665bee67a25166a2cbfe134
O block = 3a8f9c58dcfd397fede17e9dc7fd23c0667b26a847ca3bbccddf6d567358fa7c PoW = 0000085dc5a994da2d4cc161ad4110d720043c553665bee67a25166a2cbfe134
```

cryptozeny commented:
check
https://github.com/ResistancePlatform/resistance-core/compare/20c7ba33101133053bd001a84479ff341a134cba...3c80dac3618e42c90057da7ba6fa49c44cf64349

cryptozeny commented:
is it OK?

RES
![image](https://user-images.githubusercontent.com/37016180/71324327-ed619d00-2520-11ea-818e-19ac82780e0d.png)

SUGAR
![image](https://user-images.githubusercontent.com/37016180/71324333-fbafb900-2520-11ea-9979-c65ac2ee117d.png)

BTC
![image](https://user-images.githubusercontent.com/37016180/71324337-0cf8c580-2521-11ea-893d-2115bea72bcf.png)

cryptozeny reviewed (approved):
Tested ACK 8e5b110

cryptozeny commented:
TODO: remove `printf` in `block.cpp`

cryptozeny reviewed (approved):
Tested ACK 8e5b110

solardiz commented:
Also relevant is https://github.com/ResistancePlatform/resistance-core/commit/4efff3000711cc414e76cd57f2e605762b6cc339 which sets MAX_BLOCKS_IN_TRANSIT_PER_PEER to be same as MAX_HEADERS_RESULTS. Without this change, at least in Resistance the block headers download would get unnecessarily far ahead of the full blocks download, resulting in more work lost and redone in case the initial blocks download is interrupted and continued. The work loss is because we do not yet store cached PoWs on disk.

Commit:
Sugarchain: cleanup log cached PoW (Yespower)

cryptozeny commented:
great thanks to advice to us! @solardiz :smile: 

after you, i changed both. SUGAR blocktime is 5 seconds so 120x faster than BTC.
```cpp
static const int MAX_BLOCKS_IN_TRANSIT_PER_PEER = 1920; // BTC was 16 // 16*120=1920
static const unsigned int MAX_HEADERS_RESULTS = 1920; // BTC was 2000
```

but i got this error, and all banned from other nodes...
```
ERROR: headers message size = 2000
Misbehaving: 157.245.69.15:34230 peer=0 (80 -> 100) BAN THRESHOLD EXCEEDED
```
![image](https://user-images.githubusercontent.com/37016180/71327882-785a8b80-2551-11ea-915c-20d1b91850df.png)

how to fix this? :thinking:

solardiz commented:
@cryptozeny Don't change MAX_HEADERS_RESULTS. Just set MAX_BLOCKS_IN_TRANSIT_PER_PEER to some value from e.g. 200 to 2000 - experiment with this.

cryptozeny commented:
### IBD:
total: `12:30` (green: slower than before red)
download: `3:30`
sync: `9:00`
![image](https://user-images.githubusercontent.com/37016180/71356336-43881c00-25c5-11ea-8e9c-b4264be1f02b.png)

### TODO:
- [x] increase `BLOCK_DOWNLOAD_WINDOW` upto 2048 (was 1024)
```cpp
MAX_BLOCKS_IN_TRANSIT_PER_PEER = 2000;
MAX_HEADERS_RESULTS = 2000;
BLOCK_DOWNLOAD_WINDOW = 2048;
```

<details><summary>CLICK ME</summary>
<p>

![image](https://user-images.githubusercontent.com/37016180/71330792-47d71980-2572-11ea-8fe5-d9e6686965d6.png)

</p>
</details>

cryptozeny reviewed (changes requested):
those two parameters should be changed. try some experiment.
```
MAX_BLOCKS_IN_TRANSIT_PER_PEER (16)
BLOCK_DOWNLOAD_WINDOW (1024)
```

but never change `MAX_HEADERS_RESULTS` it causes [unexpected HF](https://github.com/sugarchain-project/sugarchain/pull/22#issuecomment-568306229).

> solardiz commented
Also relevant is ResistancePlatform/resistance-core@4efff30 which sets MAX_BLOCKS_IN_TRANSIT_PER_PEER to be same as MAX_HEADERS_RESULTS. Without this change, at least in Resistance the block headers download would get unnecessarily far ahead of the full blocks download, resulting in more work lost and redone in case the initial blocks download is interrupted and continued. The work loss is because we do not yet store cached PoWs on disk.

> Don't change MAX_HEADERS_RESULTS. Just set MAX_BLOCKS_IN_TRANSIT_PER_PEER to some value from e.g. 200 to 2000 - experiment with this.

cryptozeny commented:
NACK bd2d2bb :x: 
peer all lost during IBD. at first is OK (8), but later it had `0 peers` (total 3x times during download), however somehow it find another one. it works anyhow but not good.

<details><summary>CLICK ME</summary>
<p>

### IBD_2:
total `7:50`
download `4:00`
sync `3:50`
```
MAX_BLOCKS_IN_TRANSIT_PER_PEER = 2000;
BLOCK_DOWNLOAD_WINDOW = 2048;
```
![image](https://user-images.githubusercontent.com/37016180/71376988-b49bf380-2606-11ea-810a-85f04293eea9.png)
![image](https://user-images.githubusercontent.com/37016180/71366058-375f8700-25e4-11ea-8e9f-64fd6df71715.png)
![image](https://user-images.githubusercontent.com/37016180/71376777-e95b7b00-2605-11ea-95d2-2ec0ea11bfb4.png)

### IBD_3: 
total `7:30`
```
MAX_BLOCKS_IN_TRANSIT_PER_PEER = 2000;
BLOCK_DOWNLOAD_WINDOW = 4096;
```
![image](https://user-images.githubusercontent.com/37016180/71404877-11cf8d80-2677-11ea-89a8-a5185ca88067.png)

### IBD_4: (revert to ogirinal)
total `7:10`
```
MAX_BLOCKS_IN_TRANSIT_PER_PEER = 16;
BLOCK_DOWNLOAD_WINDOW = 1024;
```
![image](https://user-images.githubusercontent.com/37016180/71420133-dd7bc180-26b6-11ea-9c09-e011a37cdb46.png)
![image](https://user-images.githubusercontent.com/37016180/71420029-35fe8f00-26b6-11ea-8ffa-176e73ff03f7.png)

</p>
</details>

solardiz commented:
I suggest you keep the parameter changes separate from this PoW caching PR.

cryptozeny committed:
Merge pull request #6 from volbil/volbil-yespower_cache_5

Sugarchain: cleanup log cached PoW (Yespower)

cryptozeny reviewed (approved):
Tested ACK f3d91060fc0b1c74b202a97c80ce5fe73c8c17e0

### IBD_4: (revert to ogirinal)
total `7:10`
```
MAX_BLOCKS_IN_TRANSIT_PER_PEER = 16;
BLOCK_DOWNLOAD_WINDOW = 1024;
```
![image](https://user-images.githubusercontent.com/37016180/71420133-dd7bc180-26b6-11ea-9c09-e011a37cdb46.png)

TODO:
> I suggest you keep the parameter changes separate from this PoW caching PR.
- optimize IBD & download
- remove printf

cryptozeny commented:
https://github.com/sugarchain-project/sugarchain/issues/23
